### PR TITLE
Harden openat2 validation and /dev/fd resolution

### PR DESCRIPTION
**Pull request by abnormal749**
## Summary

Harden `openat2(2)` emulation to better match Linux semantics:

  - reject unsupported/nonzero `open_how` extension fields with `E2BIG`
  - reject oversized `open_how` structures with `E2BIG`
  - reject unknown/conflicting `how.flags` and invalid `how.mode` with `EINVAL`
  - reject `/dev/fd/N` magic-link traversal under `RESOLVE_NO_XDEV`
  - make `RESOLVE_NO_SYMLINKS` / `RESOLVE_NO_MAGICLINKS` also reject `/dev/fd/N`

Linux intentionally defines `openat2(2)` as a stricter API than `openat(2)`:
unknown or conflicting `how.flags` must fail, invalid `how.mode` must fail, nonzero unsupported extension fields must fail with `E2BIG`, and `RESOLVE_NO_XDEV` must reject mount-point crossings during path resolution.

Reference: https://man7.org/linux/man-pages/man2/openat2.2.html

## Motivation

  Before this change, elfuse accepted some invalid `open_how` inputs that Linux rejects. It also allowed a `/dev` dirfd-relative `fd/N` path to reach host `/dev/fd/N` before `RESOLVE_NO_XDEV` enforcement, which could bypass the intended guest path-resolution constraints.

## Tests

```sh
make elfuse
make -B build/test-syscall-fidelity
build/elfuse build/test-syscall-fidelity
make test-hello

  Result:

  test-syscall-fidelity: 39 passed, 0 failed, 0 skipped - PASS
 ```
 

---
## Summary by cubic
Hardened openat2(2) emulation to match Linux: strict `open_how` ABI checks and early flag/resolve validation, plus blocking fd magic-link escapes via `/proc/*/fd` and `/dev/fd`. This avoids cross-mount and magic-link bypasses and rejects invalid inputs before path lookup.

- **Bug Fixes**
  - Normalize absolute, dirfd-, and cwd-relative paths within the guest root; treat `/proc/self/fd/N`, `/proc/<self_pid>/fd/N`, and `/dev/fd/N` as fd magic-link anchors.
  - Enforce resolve policy: `RESOLVE_NO_SYMLINKS`/`RESOLVE_NO_MAGICLINKS` return `ELOOP` for those anchors; `RESOLVE_NO_XDEV` returns `EXDEV`.
  - Validate `open_how` ABI: nonzero tails fail with `E2BIG`; sizes >4096 fail with `E2BIG`.
  - Validate flags/mode before lookup: reject unknown flags, mode bits outside `07777`, mode without `O_CREAT`, `RESOLVE_BENEATH|RESOLVE_IN_ROOT`, `O_DIRECTORY|O_CREAT`, `O_PATH` misuse, and `__O_TMPFILE` usage with `EINVAL`.
  - Added flag constants (`LINUX_O_DSYNC`, `LINUX_O_SYNC`, `LINUX___O_TMPFILE`, `LINUX_O_TMPFILE`), expanded tests for `/proc/<pid>/fd` and `/dev/fd` variants, and stabilized cwd-sensitive tests using `fchdir`.

<sup>Written for commit 988364d95c452775b6a35cd262d6c3d52e582ae9. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/sysprog21/elfuse/pull/90?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/syscall/path.c [hunk 1]
```diff
@@ -902,6 +902,32 @@ static int dirfd_guest_base_path(guest_fd_t dirfd, char *out, size_t outsz)
...
+
+    if (path[0] == '/') {
+        if (path_openat2_normalize_in_root(path, normalized,
+                                           sizeof(normalized)) < 0)
+            return 0;
+        return !strncmp(normalized, "dev/fd/", 7);
+    } else {
+        char base[LINUX_PATH_MAX];
+        if (dirfd_guest_base_path(dirfd, base, sizeof(base)) < 0 ||
+            strcmp(base, "/dev") != 0)
```
Comment (jserv) [inline src/syscall/path.c]: [High] `/dev/fd` magic-link detection is incomplete: it only catches relative paths when the dirfd base equals exactly `/dev`. These all bypass `RESOLVE_NO_MAGICLINKS` / `RESOLVE_NO_SYMLINKS` / `RESOLVE_NO_XDEV` and reach host `/dev/fd`:

- `dirfd = /` + `"dev/fd/N"`
- `AT_FDCWD` with cwd `/` + `"dev/fd/N"`
- `dirfd = /tmp` + `"../dev/fd/N"`

Fix: resolve `dirfd_guest_base_path(dirfd) + path` for all relative paths (handling `.` / `..`), then match the resolved guest path against `/dev/fd/`. Mirror the structure of `resolve_proc_at_path`. The `strcmp(base, "/dev") != 0` exact-match is also brittle if `dirfd_guest_base_path` ever returns a trailing slash.

**Review by jserv — COMMENTED** [review]
Solid direction: the validation rejections match Linux semantics for the cases they cover, the new tests are tight, and the 39-case fidelity suite stays green on this branch. A few gaps and one validation/translation inconsistency worth addressing before merge; details inline.

File: src/syscall/syscall.c [hunk 2]
```diff
@@ -1362,6 +1362,64 @@ static int64_t sc_memfd_create(guest_t *g,
...
+    if (flags & ~known_flags)
+        return 0;
+    if ((flags & LINUX_O_PATH) &&
+        (flags & ~(LINUX_O_PATH | LINUX_O_CLOEXEC | LINUX_O_DIRECTORY |
+                   LINUX_O_NOFOLLOW)))
+        return 0;
+    if ((flags & LINUX___O_TMPFILE) &&
+        (flags & LINUX_O_TMPFILE) != LINUX_O_TMPFILE)
+        return 0;
+    if ((flags & LINUX_O_TMPFILE) == LINUX_O_TMPFILE &&
```
Comment (jserv) [inline src/syscall/syscall.c]: [Medium] Validation accepts `O_TMPFILE | O_WRONLY` / `O_RDWR`, but `OPEN_FLAG_MAP` in `translate.c` does not include `__O_TMPFILE`, so only `O_DIRECTORY` survives translation. The host `open()` then receives `O_DIRECTORY | O_WRONLY` and returns `EISDIR` / `ENOTDIR` rather than the Linux `O_TMPFILE` semantics (unnamed regular file). Either reject `__O_TMPFILE` up front (`-ENOTSUP` / `-EINVAL`) until `O_TMPFILE` is implemented, or extend `translate_open_flags()` to honor it. Accepting in the validator while the dispatcher silently misbehaves is the worst of both worlds.

File: src/syscall/syscall.c [hunk 3]
```diff
@@ -1374,15 +1432,22 @@ static int64_t sc_openat2(guest_t *g,
...
+    if (tail_rc < 0)
+        return tail_rc;
 
     uint64_t oflags = how[0], mode = how[1];
     uint64_t resolve = how[2];
 
+    if (!openat2_flags_valid(oflags, mode))
+        return -LINUX_EINVAL;
     if (resolve & ~(uint64_t) RESOLVE_ALL)
         return -LINUX_EINVAL;
```
Comment (jserv) [inline src/syscall/syscall.c:1457]: [Medium] Missing Linux rejection for `RESOLVE_BENEATH | RESOLVE_IN_ROOT`. `build_open_flags()` rejects this combo with `EINVAL`; here it falls through to IN_ROOT semantics (BENEATH's absolute-path `EXDEV` only trips for absolute paths, masking the issue otherwise). Add right after the unknown-resolve mask check:

```c
if ((resolve & RESOLVE_BENEATH) && (resolve & RESOLVE_IN_ROOT))
    return -LINUX_EINVAL;
```

File: src/syscall/syscall.c [hunk 4]
```diff
@@ -1362,6 +1362,64 @@ static int64_t sc_memfd_create(guest_t *g,
...
+        LINUX_O_LARGEFILE | LINUX_O_NOATIME | LINUX_O_CLOEXEC | LINUX_O_SYNC |
+        LINUX_O_PATH | LINUX___O_TMPFILE;
+
+    if (flags & ~known_flags)
+        return 0;
+    if ((flags & LINUX_O_PATH) &&
+        (flags & ~(LINUX_O_PATH | LINUX_O_CLOEXEC | LINUX_O_DIRECTORY |
+                   LINUX_O_NOFOLLOW)))
+        return 0;
+    if ((flags & LINUX___O_TMPFILE) &&
```
Comment (jserv) [inline src/syscall/syscall.c]: [Medium] `O_DIRECTORY | O_CREAT` is not rejected here. Linux `build_open_flags()` returns `EINVAL` up front; this validator forwards the combo, and `sys_openat_path()` then produces host-dependent errno. Since `LINUX_O_TMPFILE` already includes `LINUX_O_DIRECTORY`, the same check also closes `O_TMPFILE | O_CREAT` (which Linux's `O_TMPFILE_MASK` rejects):

```c
if ((flags & (LINUX_O_DIRECTORY | LINUX_O_CREAT)) ==
    (LINUX_O_DIRECTORY | LINUX_O_CREAT))
    return 0;
```

File: src/syscall/syscall.c [hunk 5]
```diff
@@ -1443,14 +1509,17 @@ static int64_t sc_openat2(guest_t *g,
...
-             * /proc/<pid>/fd/N with an explicit pid. A NO_XDEV resolution
-             * landing on either is still a real cross, but the precheck has
-             * no normalization path for /dev or for explicit-pid procfs
-             * anchors yet, so the bypass remains.
+             * /dev/fd/N is the same kind of magic-link escape from the
+             * guest's perspective; catch both absolute and /dev-dirfd
+             * relative forms before the fallback host openat path can reach
+             * macOS /dev/fd.
+             *
+             * Still not detected: /proc/<pid>/fd/N with an explicit pid
```
Comment (jserv) [inline src/syscall/syscall.c]: [Medium] Worth tracking as a real bypass, not just a documentation note: `path_openat2_is_proc_magiclink` only matches normalized `/proc/self/fd/`, so `/proc/1234/fd/N` (with `1234 == getpid()`) traverses the same magic links and currently slips past `RESOLVE_NO_MAGICLINKS` / `RESOLVE_NO_XDEV`. Acknowledged here as a known limitation, but the fix scope is small: normalize self-pid forms in `proc_seed_absolute_path` before the prefix check.

File: tests/test-syscall-fidelity.c [hunk 6]
```diff
@@ -816,6 +871,36 @@ static void test_openat2_resolve_no_xdev_rejects_proc_fd_magiclink(void)
     EXPECT_TRUE(errno == EXDEV, "wrong errno");
 }
 
+static void test_openat2_resolve_no_xdev_rejects_dev_fd_magiclink(void)
```
Comment (jserv) [inline tests/test-syscall-fidelity.c:985]: [Low] Coverage is narrow. The new `/dev/fd` case only exercises `dirfd=/dev` with `RESOLVE_NO_XDEV`. Suggested additions to lock in the boundary once the helper is broadened:

- absolute `/dev/fd/N` under `RESOLVE_NO_MAGICLINKS` and `RESOLVE_NO_SYMLINKS` (expect `ELOOP`)
- `dirfd=/` + `"dev/fd/N"` under each of the three resolve flags
- `AT_FDCWD` with cwd `/` + relative form
- `O_TMPFILE | O_RDONLY` and `O_DIRECTORY | O_CREAT` (expect `EINVAL`) once those validations land

**Comment by abnormal749** [comment]
commit 1: Fix openat2 magic-link path normalization
- centralize fd magic-link detection through `path_openat2_is_fd_magiclink_anchor()`
- broaden `/dev/fd` relative path detection
- handle absolute `/dev/fd/N`
- handle `dirfd=/` + `dev/fd/N`
- handle `AT_FDCWD` with cwd `/` + `dev/fd/N`
- handle normalized `/dev` paths such as `shm/../fd/N`
- handle `/proc/<pid>/fd/N` where `pid == getpid()`
- keep the extra check limited to openat2 resolve-flag prechecks

I also consolidated the `/dev/fd`, `/proc/self/fd`, and `/proc/<self_pid>/fd` checks into `path_openat2_is_fd_magiclink_anchor()`. The intent is to treat them as the same openat2 policy concept: fd magic-link anchors. The helper only performs lexical normalization against the dirfd/cwd base and does not follow symlinks or change general proc/dev path handling. This keeps the extra work limited to resolve-flag prechecks while avoiding scattered special cases.

commit 2: Tighten openat2 flag validation
- reject `RESOLVE_BENEATH | RESOLVE_IN_ROOT`
- reject `O_DIRECTORY | O_CREAT`
- reject `__O_TMPFILE` until Linux `O_TMPFILE` unnamed-file semantics are actually implemented
- avoid accepting flags that `translate_open_flags()` cannot faithfully preserve

commit 3: Add fidelity tests for reviewed edge cases
- `RESOLVE_BENEATH | RESOLVE_IN_ROOT` -> `EINVAL`
- `O_DIRECTORY | O_CREAT` -> `EINVAL`
- `O_TMPFILE | O_RDONLY` -> `EINVAL`
- `/proc/<self_pid>/fd/N` absolute and `dirfd=/proc` relative forms under `NO_MAGICLINKS` / `NO_SYMLINKS` / `NO_XDEV`
- `/dev/fd/N` absolute, `dirfd=/ + dev/fd/N`, cwd `/ + dev/fd/N`, and `dirfd=/dev + shm/../fd/N` under the same three resolve policies

One detail: I did not keep the literal `dirfd=/tmp + ../dev/fd/N` case because under macOS/elfuse `open("/tmp")` resolves to `/private/tmp`, so `../dev/fd` becomes `/private/dev/fd` and correctly returns `ENOENT`. I used `dirfd=/dev + shm/../fd/N` to cover the same `..` lexical normalization into `/dev/fd/N` without depending on macOS `/tmp` symlink behavior.

Thank you for the detailed review. I tried to keep the fixes narrowly scoped, separate the behavior changes from the tests, and make the validation path closer to Linux semantics while avoiding unsupported `O_TMPFILE` behavior for now.

**Comment by jserv** [comment]
> 等級	意見	修復	Commit

Please don't create tables like this. We're here to discuss and improve the software together, not to act as task trackers whose sole purpose is to clear action items and close pending items without genuine interaction.

**Comment by abnormal749** [comment]
Understood. I removed the checklist-style table and will keep the discussion focused on the reasoning and tradeoffs instead.

File: src/syscall/path.c [hunk 7]
```diff
@@ -902,6 +873,50 @@ static int dirfd_guest_base_path(guest_fd_t dirfd, char *out, size_t outsz)
     return 0;
 }
 
+static int normalized_proc_self_fd_anchor(const char *path)
```
Comment (jserv) [inline src/syscall/path.c]: Use `bool` when possible.

File: src/syscall/path.h [hunk 8]
```diff
@@ -65,7 +65,7 @@ const char *path_resolve_sysroot_create_path(const char *path,
 
 int path_openat2_stays_beneath(const char *path, bool clamp_at_root);
 int path_openat2_normalize_in_root(const char *path, char *out, size_t outsz);
-int path_openat2_is_proc_magiclink(guest_fd_t dirfd, const char *path);
+int path_openat2_is_fd_magiclink_anchor(guest_fd_t dirfd, const char *path);
```
Comment (jserv) [inline src/syscall/path.h]: Use `bool` when possible.

File: src/syscall/syscall.c [hunk 9]
```diff
@@ -1362,6 +1362,62 @@ static int64_t sc_memfd_create(guest_t *g,
...
+        for (size_t i = 0; i < chunk; i++) {
+            if (tail[i] != 0)
+                return -LINUX_E2BIG;
+        }
+        off += chunk;
+    }
+    return 0;
+}
+
+static int openat2_flags_valid(uint64_t flags, uint64_t mode)
```
Comment (jserv) [inline src/syscall/syscall.c]: Use `bool` when possible.

File: tests/test-syscall-fidelity.c [hunk 10]
```diff
@@ -184,6 +231,83 @@ static void test_openat2_basic(void)
     PASS();
 }
 
+static void test_openat2_rejects_nonzero_how_extension(void)
+{
+    TEST("openat2 rejects nonzero open_how extension");
+    struct {
+        struct open_how how;
+        unsigned long long extra;
+    } ext = {.how = {.flags = O_RDONLY, .mode = 0, .resolve = 0}, .extra = 1};
```
Comment (jserv) [inline tests/test-syscall-fidelity.c]: Add comma right after `.extra = 1`, so that clang-format can indent better.

File: src/syscall/path.c [hunk 11]
```diff
@@ -902,6 +873,50 @@ static int dirfd_guest_base_path(guest_fd_t dirfd, char *out, size_t outsz)
...
+{
+    if (!strncmp(path, "proc/self/fd/", 13))
+        return 1;
+    if (strncmp(path, "proc/", 5))
+        return 0;
+
+    char *endp;
+    errno = 0;
+    long long pid = strtoll(path + 5, &endp, 10);
+    if (endp == path + 5 || errno == ERANGE ||
```
Comment (jserv) [inline src/syscall/path.c]: Explain `5`.

File: src/syscall/path.c [hunk 12]
```diff
@@ -902,6 +873,50 @@ static int dirfd_guest_base_path(guest_fd_t dirfd, char *out, size_t outsz)
...
+        pid != (long long) proc_get_pid())
+        return 0;
+    return !strncmp(endp, "/fd/", 4);
+}
+
+int path_openat2_is_fd_magiclink_anchor(guest_fd_t dirfd, const char *path)
```
Comment (jserv) [inline src/syscall/path.c]: Use `bool` when possible.

File: src/syscall/syscall.c [hunk 13]
```diff
@@ -1362,6 +1362,62 @@ static int64_t sc_memfd_create(guest_t *g,
     (RESOLVE_NO_XDEV | RESOLVE_NO_MAGICLINKS | RESOLVE_NO_SYMLINKS | \
      RESOLVE_BENEATH | RESOLVE_IN_ROOT | RESOLVE_CACHED)
 
+#define OPEN_HOW_SIZE_VER0 24
+#define OPEN_HOW_MAX_SIZE 4096
```
Comment (jserv) [inline src/syscall/syscall.c:1373]: As `open_how` is Linux specific, add proper comments to address the behavior. 
See https://man7.org/linux/man-pages/man2/open_how.2type.html

**Review by jserv — CHANGES_REQUESTED** [review]
Squash the commits and revise the commit messages to comply with the guidelines at https://cbea.ms/git-commit/

**Comment by abnormal749** [comment]
I have pushed the cleaned branch with the `openat2` work and squashed into one commit on top of the latest upstream/main, after the PTY ioctl changes.

The latest review comments are addressed as well: `bool` is used for boolean helpers, `open_how` now has a short comment explaining the Linux extensible ABI size behavior, and the self-pid /proc fd parser uses a named `pid_start` pointer instead of the raw path + 5 offset.

Local validation passes, including format/security/shellcheck/cppcheck, clang-tidy, scan-build, and the syscall fidelity test.

**Comment by jserv** [comment]
> I have pushed the cleaned branch with the `openat2` work and squashed into one commit on top of the latest upstream/main, after the PTY ioctl changes.

You don't need to summarize your changes, as we can always review them using Git. Instead, simply click "Resolve conversation" via GitHub web when the issue is addressed, or leave comments wherever you have concerns. Please avoid imitating the workflow or actions of AI agents.

**Comment by jserv** [comment]
Thank @abnormal749 for contributing!
